experimental/experimental_examples/cpp_api/taskgraph_filtering.cc

```diff
@@ -0,0 +1,288 @@
+/**
+ * @file    taskgraph_filtering.cc
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2023 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This example demonstrates basic functionality of task graph APIs.
+ * It reads an array, compresses the data, decompresses, writes data into
+ * a new array, compares the new array data with original data.
+ *
+ *
+ */
+
+#include <iostream>
+
+#include <tiledb/tiledb>
+#include "experimental/tiledb/common/dag/edge/edge.h"
+#include "experimental/tiledb/common/dag/execution/bountiful.h"
+#include "experimental/tiledb/common/dag/execution/throw_catch.h"
+#include "experimental/tiledb/common/dag/graph/taskgraph.h"
+#include "experimental/tiledb/common/dag/nodes/segmented_nodes.h"
+
+#include <bzlib.h>
+
+using namespace tiledb;
+using namespace tiledb::common;
+
+template <class T>
+using input_info =
+    std::tuple<uint32_t, uint32_t, std::shared_ptr<std::vector<T>>>;
+
+class chunk_generator {
+  inline static std::atomic<uint32_t> begin_{0};
+  uint32_t array_size_;
+  uint32_t chunk_size_;
+
+ public:
+  chunk_generator(uint32_t array_size, uint32_t chunk_size)
+      : array_size_(array_size)
+      , chunk_size_(chunk_size) {
+  }
+
+  void reset() {
+    begin_.store(0);
+  }
+
+  uint32_t operator()(std::stop_source& stop_source) {
+    if (begin_ + chunk_size_ > array_size_) {
+      stop_source.request_stop();
+      // TODO: Unclear how graph works here, does it
+      // pulsate one more time or it dies including
+      // this current operation?
+    }
+
+    uint32_t rv = begin_;
+    begin_ += chunk_size_;
+    return rv;
+  }
+};
+
+void create_array(Context& ctx, const std::string& array_name) {
+  Domain domain(ctx);
```

Comment:
Our other examples include more comments. Let's try to follow the same commenting style. We should strive for more consistency across all examples. People read examples to try to make sense of our API.